pkg/agent/plugin/nodeattestor/v0_test.go

```diff
@@ -0,0 +1,259 @@
+			streamImpl:    &fakeStream{expectData: attestationData, challenges: challenges("echo")},
+			expectCode:    codes.InvalidArgument,
+			expectMessage: "stream received invalid challenge response",
+		},
+		{
+			test:          "plugin response with empty challenge response",
+			pluginImpl:    &fakePluginV0{attestationData: &attestationData, challengeResponses: challengeResponses("echo", "")},
+			streamImpl:    &fakeStream{expectData: attestationData, challenges: challenges("echo")},
+			expectCode:    codes.Internal,
+			expectMessage: "test plugin: plugin response missing challenge response",
+		},
+	} {
+		tt := tt
+		t.Run(tt.test, func(t *testing.T) {
+			nodeattestor := loadFakePluginV0(t, tt.pluginImpl)
+			err := nodeattestor.Attest(context.Background(), tt.streamImpl)
+			if tt.expectCode != codes.OK {
+				spiretest.RequireGRPCStatus(t, err, tt.expectCode, tt.expectMessage)
+				return
+			}
+			require.NoError(t, err)
+		})
+	}
+
+}
+
+func loadFakePluginV0(t *testing.T, fake *fakePluginV0) nodeattestor.NodeAttestor {
+	server := nodeattestorv0.PluginServer(fake)
+
+	client := struct {
+		catalog.PluginInfo
+		nodeattestorv0.NodeAttestor
+	}{}
+
+	spiretest.LoadPlugin(t, catalog.MakePlugin("test", server), &client)
+	return nodeattestor.V0(client, client)
+}
+
+type fakePluginV0 struct {
+	nodeattestorv0.UnimplementedNodeAttestorServer
+
+	closeStream        bool
+	attestationData    *nodeattestor.AttestationData
+	attestationDataErr error
+
+	challengeResponses   map[string]string
+	challengeResponseErr error
+}
+
+func (plugin *fakePluginV0) FetchAttestationData(stream nodeattestorv0.NodeAttestor_FetchAttestationDataServer) error {
+	if plugin.closeStream {
+		return nil
+	}
+	if plugin.attestationDataErr != nil {
+		return plugin.attestationDataErr
+	}
+	if err := stream.Send(&nodeattestorv0.FetchAttestationDataResponse{
+		AttestationData: v0AttestationData(plugin.attestationData),
+	}); err != nil {
+		return err
+	}
+
+	for len(plugin.challengeResponses) > 0 {
+		req, err := stream.Recv()
+		if err != nil {
+			return err
+		}
+		challenge := string(req.Challenge)
+		if plugin.challengeResponseErr != nil {
+			return plugin.challengeResponseErr
+		}
+		response, ok := plugin.challengeResponses[challenge]
+		if !ok {
+			return fmt.Errorf("test not configured to handle challenge %q", challenge)
+		}
+		delete(plugin.challengeResponses, challenge)
+		if err := stream.Send(&nodeattestorv0.FetchAttestationDataResponse{
+			Response: []byte(response),
+		}); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (plugin *fakePluginV0) Configure(context.Context, *spi.ConfigureRequest) (*spi.ConfigureResponse, error) {
+	return &spi.ConfigureResponse{}, nil
+}
+
+func (plugin *fakePluginV0) GetPluginInfo(context.Context, *spi.GetPluginInfoRequest) (*spi.GetPluginInfoResponse, error) {
+	return &spi.GetPluginInfoResponse{}, nil
+}
+
+type fakeStream struct {
```

Comment:
This fake implementation is only here temporarily. When v1 is introduced, it will probably be moved elsewhere.